src/web_client.c

```diff
@@ -2047,95 +2108,123 @@ void *web_client_main(void *ptr)
 	if(pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL) != 0)
 		error("Cannot set pthread cancel state to ENABLE.");
 
-	struct timeval tv;
 	struct web_client *w = ptr;
-	int retval;
-	fd_set ifds, ofds, efds;
-	int fdmax = 0;
+	struct pollfd fds[2], *ifd, *ofd;
+	int retval, fdmax = 0, timeout;
 
 	log_access("%llu: %s port %s connected on thread task id %d", w->id, w->client_ip, w->client_port, gettid());
 
 	for(;;) {
-		FD_ZERO (&ifds);
-		FD_ZERO (&ofds);
-		FD_ZERO (&efds);
-
-		FD_SET(w->ifd, &efds);
-
-		if(w->ifd != w->ofd)
-			FD_SET(w->ofd, &efds);
-
-		if (w->wait_receive) {
-			FD_SET(w->ifd, &ifds);
-			if(w->ifd > fdmax) fdmax = w->ifd;
+		if(unlikely(w->dead)) {
+			debug(D_WEB_CLIENT, "%llu: client is dead.", w->id);
+			break;
+		}
+		else if(unlikely(!w->wait_receive && !w->wait_send)) {
+			debug(D_WEB_CLIENT, "%llu: client is not set for neither receiving nor sending data.");
+			break;
 		}
 
-		if (w->wait_send) {
-			FD_SET(w->ofd, &ofds);
-			if(w->ofd > fdmax) fdmax = w->ofd;
+		if(unlikely(w->ifd < 0 || w->ofd < 0)) {
+			error("%llu: invalid file descriptor, ifd = %d, ofd = %d (required 0 <= fd", w->id, w->ifd, w->ofd);
+			break;
 		}
 
-		tv.tv_sec = web_client_timeout;
-		tv.tv_usec = 0;
+		if(w->ifd == w->ofd) {
+			fds[0].fd = w->ifd;
+			fds[0].events = 0;
+			fds[0].revents = 0;
 
-		debug(D_WEB_CLIENT, "%llu: Waiting socket async I/O for %s %s", w->id, w->wait_receive?"INPUT":"", w->wait_send?"OUTPUT":"");
-		retval = select(fdmax+1, &ifds, &ofds, &efds, &tv);
+			if(w->wait_receive) fds[0].events += POLLIN;
```

Comment:
You are right. This was a fail in my thoughts. Anyways I think `|` is more clean to read. Thank you for changing it.